ucey-star/ID8
# Google authentication end-to-end

## Description
**Note**: This PR has many files since I rebased the new main on this branch. This is because I had created it a while back and given the PRs we merged yesterday we had very many new changes on main hence the need to rebase this branch since it was created from an older version of main.

This PR introduces end-to-end functionality for google Sign In. There are quite a lot of files however here is an overview of how they are structured and what they do.
1. I have put all auth related files inside the `auth` folder. Inside this folder we have the signup and login pages as well as the `actions.tsx` which is has all the auth handlers for signIn with google ,email and password, as well as functionality to Sign Out. 
2. We have two `middleware.tsx` files on in the api folder and another in the src folder. This two files are copied directly from the supabase documentation and are used to verify on every route whether a user is authenticated and authorized to view information on a particular page. The `middleware.tsx` also is the same file that redirects users to the login page when they are not authenticated as well as new users. 

3. I have moved all the folders to the components folder outside app directory as they are not server side components. I intend to delete the _components folder in another PR as it is quite confusing. Additionally, this change was also done due to the me putting all auth related files in the same folder. To be clear my decision to move all auth related things into one folder is due to following the supabase google auth documentation. I am aware this is not necessary however this made it easier for me to follow the tutorial and troubleshoot issues. 

## Type of Change


## Testing Instructions
1. Check into the branch
2. You will be automatically be redirected to the login page. If this occurs then you are sure the `middleware.tsx` files are working as expected. 
3. Login/signUp and you should be be directed to the / directory if the login is successful.

## Screenshots (if applicable)

No UI changes were made.
